core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -755,104 +769,115 @@ private[spark] class BlockManager(
-          putBlockInfo.markFailure()
-          logWarning(s"Putting block $blockId failed")
         }
       }
-    }
-    logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
+      logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
 
-    // Either we're storing bytes and we asynchronously started replication, or we're storing
-    // values and need to serialize and replicate them now:
-    if (putLevel.replication > 1) {
-      data match {
-        case ByteBufferValues(bytes) =>
-          if (replicationFuture != null) {
-            Await.ready(replicationFuture, Duration.Inf)
-          }
-        case _ =>
-          val remoteStartTime = System.currentTimeMillis
-          // Serialize the block if not already done
-          if (bytesAfterPut == null) {
-            if (valuesAfterPut == null) {
-              throw new SparkException(
-                "Underlying put returned neither an Iterator nor bytes! This shouldn't happen.")
+      // Either we're storing bytes and we asynchronously started replication, or we're storing
+      // values and need to serialize and replicate them now:
+      if (putLevel.replication > 1) {
+        data match {
+          case ByteBufferValues(bytes) =>
+            if (replicationFuture != null) {
+              Await.ready(replicationFuture, Duration.Inf)
             }
-            bytesAfterPut = dataSerialize(blockId, valuesAfterPut)
-          }
-          replicate(blockId, bytesAfterPut, putLevel)
-          logDebug("Put block %s remotely took %s"
-            .format(blockId, Utils.getUsedTimeMs(remoteStartTime)))
+          case _ =>
+            val remoteStartTime = System.currentTimeMillis
+            // Serialize the block if not already done
+            if (bytesAfterPut == null) {
+              if (valuesAfterPut == null) {
+                throw new SparkException(
+                  "Underlying put returned neither an Iterator nor bytes! This shouldn't happen.")
+              }
+              bytesAfterPut = dataSerialize(blockId, valuesAfterPut)
+            }
+            replicate(blockId, bytesAfterPut, putLevel)
+            logDebug("Put block %s remotely took %s"
+              .format(blockId, Utils.getUsedTimeMs(remoteStartTime)))
+        }
       }
-    }
 
-    BlockManager.dispose(bytesAfterPut)
+    } finally {
+      BlockManager.dispose(bytesAfterPut)
+      // this is to clean up the byte buffer from the *input* (as opposed to the line above, which
+      // disposes the byte buffer that is the *result* of the put).  We might have turned that byte
+      // buffer into an iterator of values, in which case the input ByteBuffer should be disposed.
+      // It will automatically get disposed when we get to the end of the iterator, but we might
+      // never even try to get to the end, or there might be an exception along the way
+      resourceCleaner.doCleanup()
```

Comment:
repeating my previous comment, since it still applies despite the changed diff:

this is probably the only case that is really unusual, and worth getting somebody else to look at. A call to `BlockManager.doPut` might try to deserialize a ByteBuffer its given (if storageLevel is memory deserialized), and then it might read to the end of that Iterator (if there is enough memory to read to the end of the iterator, or if the storage levels allows dropping to disk). If both of those are true, the old code would dispose the input byte buffer. If not, then it never would.

So the issue here is not so much that we're worried about exceptions from user code (since the iterator is never exposed to the user), but just being consistent on whether or not we dispose the input byte buffer.